### PR TITLE
[29_6] add pdf format and conversion mechanism betwen tm and pdf

### PR DESCRIPTION
**Pull request by tangdouer1005**
## Why you open this Pull Request?
Previously, there were many drawbacks to opening PDFs through truncation, such as not being able to save properly, and detaching attachments may overwrite other files. These problems can be perfectly solved by adding pdf format and its conversion mechanism

## What work have you done in the current Pull Request?
add pdf format and conversion mechanism betwen tm and pdf

### CODE REVIEW

**Comment by darcy-shen** [comment]
Please update the pr description properly.

File: plugins/pdf/progs/init-pdf.scm [hunk 1]
```diff
@@ -0,0 +1,53 @@
+(texmacs-module (pdf-format))
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:1]: 最上面没有 LICENSE信息

File: plugins/pdf/progs/init-pdf.scm [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+
+(define (texmacs->pdf x . opts)
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:4]: scm代码的格式注意一下

File: src/Scheme/Plugins/glue_pdf.lua [hunk 3]
```diff
@@ -33,6 +33,42 @@ function main()
                     "url"
                 }
             },
+            {
+                scm_name = "make-attachments",
```
Comment (darcy-shen) [inline src/Scheme/Plugins/glue_pdf.lua:37]: 这里的命名最好都采用 pdf 开头

Comment (darcy-shen) [reply src/Scheme/Plugins/glue_pdf.lua:37]: 比如 pdf-make-attachments

File: src/Scheme/Plugins/glue_pdf.lua [hunk 4]
```diff
@@ -33,6 +33,42 @@ function main()
...
+                cpp_name = "pdf_hummus_make_attachments",
+                ret_type = "bool",
+                arg_list = {
+                    "url",
+                    "array_url",
+                    "url"
+                }
+            },
+            {
+                scm_name = "get-linked-file-paths",
```
Comment (darcy-shen) [inline src/Scheme/Plugins/glue_pdf.lua:47]: pdf-get-linked-file-paths

File: src/Scheme/Plugins/glue_pdf.lua (hunk 3 above)
Comment (darcy-shen) [reply src/Scheme/Plugins/glue_pdf.lua:37]: 比如url相关的函数，基本上都是 url-xxx

File: src/Scheme/Plugins/glue_pdf.lua [hunk 5]
```diff
@@ -33,6 +33,42 @@ function main()
...
+                cpp_name = "get_linked_file_paths",
+                ret_type = "array_url",
+                arg_list = {
+                    "tree",
+                    "url"
+                }
+            },
+            {
+                scm_name = "replace-with-relative-path",
```
Comment (darcy-shen) [inline src/Scheme/Plugins/glue_pdf.lua:56]: 同上

File: src/Scheme/Plugins/glue_pdf.lua [hunk 6]
```diff
@@ -33,6 +33,42 @@ function main()
...
+                cpp_name = "replace_with_relative_path",
+                ret_type = "tree",
+                arg_list = {
+                    "tree",
+                    "url"
+                }
+            },
+            {
+                scm_name = "get-main-tm",
```
Comment (darcy-shen) [inline src/Scheme/Plugins/glue_pdf.lua:65]: pdf-get-attached-tm?

File: plugins/pdf/progs/init-pdf.scm [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+				(tem-pdf-out (open-output-file (url->string tem-pdf))))
+	(display* tem-dir "\n")
+	(display x tem-pdf-out)
+	(close-output-port tem-pdf-out)
+	(if (extract-attachments tem-pdf)
+		()(display* "extract-attachments return false\n"))
+	(set! tem-tm (url-relative tem-tm (get-main-tm tem-pdf)))
+
+	(if (url-exists? tem-tm)
+				(display* tem-tm " exist\n")
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:44]: remove `display*`

File: plugins/pdf/progs/init-pdf.scm [hunk 8]
```diff
@@ -0,0 +1,65 @@
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; MODULE      : init-pdf.scm
+;; DESCRIPTION : setup pdf converters
+;; COPYRIGHT   : (C) 2023  tandouer
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:5]: tangdouer

File: plugins/pdf/progs/init-pdf.scm [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+		(buffer-set-master buf cur)
+		(switch-to-buffer buf)
+		(load-buffer tem-tm)
+		(set-drd cur)
+		(dynamic-make-slides)
+		(print-to-file tem-pdf)
+		(switch-to-buffer cur)
+		(buffer-close buf))
+	;;attachment-list need be generated
+	(if(pdf-make-attachments tem-pdf url-list tem-pdf)
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:36]: if 后面少一个空格

File: plugins/pdf/progs/init-pdf.scm [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+		(display* "pdf-make-attachments return true\n")
+		(display* "pdf-make-attachments return false\n"))
+	(string-load tem-pdf)
+	))
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:40]: 这个应该放在 39行最后面

File: plugins/pdf/progs/init-pdf.scm [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+(define (pdf->texmacs x . opts)
+	(let* ((tem-dir (url-temp-dir))
+			(tem-pdf (string->url (string-append (url->string tem-dir) "/tem.pdf")))
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:44]: 代码格式化

File: src/Edit/Editor/edit_main.cpp [hunk 12]
```diff
@@ -394,78 +360,18 @@ edit_main_rep::attach_doc_to_exported_pdf (url pdf_name) {
...
-              string name= basename (style_url);
-              la[i][0]->label= name;
-            }
-          }
-        }
-        else st= st * la[i];
-      }
-    }
-  }
+  array<url>  tm_and_linked_file = get_linked_file_paths(new_t, cur_u);
```
Comment (darcy-shen) [inline src/Edit/Editor/edit_main.cpp:363]: 这块代码需要放到scheme里面去实现，因为现在这样实现会导致 src/Edit/Editor 依赖 Plugin/PDF

File: src/Edit/Editor/edit_main.cpp [hunk 13]
```diff
@@ -394,78 +360,18 @@ edit_main_rep::attach_doc_to_exported_pdf (url pdf_name) {
...
+  new_t = replace_with_relative_path(new_t, cur_u);
+  array<url> attachments;
+  attachments = append(tm_and_linked_file, attachments);
+  attachments = append(new_u, attachments);
+  cout << "attachments -> " << attachments << LF;
```
Comment (darcy-shen) [inline src/Edit/Editor/edit_main.cpp:368]: 移除cout

File: src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp [hunk 14]
```diff
@@ -21,11 +21,16 @@
...
+#include "analyze.hpp"
+#include "file.hpp"
+#include "new_buffer.hpp"
 #include "pdf_hummus_extract_attachment.hpp"
 #include "tm_debug.hpp"
+#include "tm_url.hpp"
+#include "tree_helper.hpp"
 
+#include "sys_utils.hpp"
 bool
```
Comment (darcy-shen) [inline src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp:34]: include和下面代码需要空格隔开

File: plugins/pdf/progs/init-pdf.scm [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    (buffer-set-master buf cur)
+    (switch-to-buffer buf)
+    (load-buffer tem-tm)
+    (set-drd cur)
+    (dynamic-make-slides)
+    (print-to-file tem-pdf)
+    (switch-to-buffer cur)
+    (buffer-close buf))
+  ;;attachment-list need be generated
+  (if(pdf-make-attachments tem-pdf url-list tem-pdf)
```
Comment (darcy-shen) [inline plugins/pdf/progs/init-pdf.scm:36]: if后面没有空格

File: src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp [hunk 16]
```diff
@@ -155,5 +160,194 @@ extract_attachments_from_pdf (url pdf_path, list<url>& names) {
...
+    url sty_u= descendance ("$TEXMACS_PATH/styles");
+    declare_style (sty_u);
+    sty_u= descendance ("$TEXMACS_PATH/packages");
+    declare_style (sty_u);
+  }
+  return internal_styles->contains (style);
+}
+
+array<url>
+get_linked_file_paths (tree t, url path) {
```
Comment (darcy-shen) [inline src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp:202]: 这个方法的if else嵌套太多了，可读性太差，需要重构一下

File: src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp [hunk 17]
```diff
@@ -155,5 +160,194 @@ extract_attachments_from_pdf (url pdf_path, list<url>& names) {
...
+        else st= st * la[i];
+      }
+    }
+  }
+
+  return tm_and_linked_file;
+}
+
+tree
+replace_with_relative_path (tree t, url path) {
```
Comment (darcy-shen) [inline src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp:270]: 这个也有上面的问题

Comment (tangdouer1005) [reply src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp:270]: ok，或许可以拆分成小函数？主体只负责遍历，对style和image树的处理放在辅助函数里？

Comment (darcy-shen) [reply src/Plugins/Pdf/pdf_hummus_extract_attachment.cpp:270]: 对 可以拆分，后面再做吧，我们先merge吧

**Review by darcy-shen — APPROVED** [review]
LGTM now. Please improve the code quality in the follow-up pull requests.

**Pull request merged**